cleanrl/ppo_mujoco_envpool_xla_jax.py

```diff
@@ -0,0 +1,679 @@
+        logits, stdlog = actor.apply(agent_state.params.actor_params, next_obs)
+        key, subkey = jax.random.split(key)
+        u = jax.random.normal(subkey, shape=logits.shape)
+        action = logits + u * jnp.exp(stdlog)
+        var = jnp.exp(2 * stdlog)
+        logprob = (-((action - logits) ** 2) / (2 * var) - stdlog - jnp.log(jnp.sqrt(2 * jnp.pi))).sum(-1)
+        value = critic.apply(agent_state.params.critic_params, next_obs)
+        return action, logprob, value.squeeze(1), key
+
+    @jax.jit
+    def get_action_and_value2(
+        params: flax.core.FrozenDict,
+        x: np.ndarray,
+        action: np.ndarray,
+    ):
+        """calculate value, logprob of supplied `action`, and entropy"""
+        logits, stdlog = actor.apply(params.actor_params, x)
+        var = jnp.exp(2 * stdlog)
+        logprob = (-((action - logits) ** 2) / (2 * var) - stdlog - jnp.log(jnp.sqrt(2 * jnp.pi))).sum(-1)
+        entropy = (2 * stdlog + jnp.log(2 * jnp.pi) + 1) / 2
+        value = critic.apply(params.critic_params, x).squeeze()
+        return logprob, entropy.sum(-1), value
+
+    def compute_gae_once(carry, inp, gamma, gae_lambda):
+        advantages = carry
+        nextdone, nexttruncated, nextvalues, curvalues, reward = inp
+        nextnonterminal = (1.0 - nextdone) * (1.0 - nexttruncated)
+
+        delta = reward + gamma * nextvalues - curvalues
+        advantages = delta + gamma * gae_lambda * nextnonterminal * advantages
+        return advantages, advantages
+
+    compute_gae_once = partial(compute_gae_once, gamma=args.gamma, gae_lambda=args.gae_lambda)
+
+    @jax.jit
+    def compute_gae(
+        agent_state: TrainState,
+        next_obs: np.ndarray,
+        next_done: np.ndarray,
+        next_truncated: np.ndarray,
+        storage: Storage,
+    ):
+        values = critic.apply(
+            agent_state.params.critic_params,
+            jnp.concatenate([storage.obs, next_obs[None, :]], axis=0),
+        ).squeeze()
+        if args.rew_norm:
+            values = values * jnp.sqrt(agent_state.ret_rms.var).astype(jnp.float32)
+        advantages = jnp.zeros((args.num_envs,))
+        dones = jnp.concatenate([storage.dones, next_done[None, :]], axis=0)
+        truncated = jnp.concatenate([storage.truncated, next_truncated[None, :]], axis=0)
+        _, advantages = jax.lax.scan(
+            compute_gae_once,
+            advantages,
+            (
+                dones[1:],
+                truncated[1:],
+                values[1:] * (1.0 - dones[1:]),
+                values[:-1],
+                storage.rewards,
+            ),
+            reverse=True,
+        )
+        returns = advantages + values[:-1]
+        if args.rew_norm:
+            returns = (returns / jnp.sqrt(agent_state.ret_rms.var + 10e-8)).astype(jnp.float32)
+            agent_state = agent_state.replace(ret_rms=agent_state.ret_rms.update(returns.flatten()))
```

Comment:
Interesting. Going through the logic here I realize this reward normalization is quite different from the original implementation, which may or may not be fine but it's worth pointing out the difference.

The original implementation does a "forward discounted return" and normalizes the reward on a per-step basis. 
<img width="1118" alt="image" src="https://user-images.githubusercontent.com/5555347/215813439-de92bae3-1b36-4c02-b29a-bcbd4f9378a1.png">
See [source1](https://github.com/openai/baselines/blob/ea25b9e8b234e6ee1bca43083f8f3cf974143998/baselines/common/vec_env/vec_normalize.py#L28) and [source2](https://iclr-blog-track.github.io/2022/03/25/ppo-implementation-details/)

Here what happens is that the returns are normalized only after the rollout phase, so the rewards themselves are not normalized.


It doesn't feel like these two approaches are equivalent. If you want to get to the bottom of this, it might be worth it to conduct an empirical study comparing this implementation and https://github.com/vwxyzjn/cleanrl/pull/348.

Both approaches seem fine to me, but we should document if we choose the current apporach.

Reply:
I also noticed there is no observation normalization. Is this intended?